Build fails with Apple Clang 12 (clang-1200.0.32.2) due to use of reference types

Clang build fails due to the -WError flag, resulting in -Wrange-loop-analysis warnings being shown as errors. 
The issue arises in multiple files due to the use of `auto` instead of `auto&` since the compiler expects a non reference type for places where one iterates over a loop in such a way that creates a proxy/copy of the elements instead of a reference to the elements themselves.
e.g.

`
src/util/graph.h:635:20: error: loop variable 'succ' of type 'const std::__1::pair<const unsigned long, empty_edget>' creates a
      copy from type 'const std::__1::pair<const unsigned long, empty_edget>' [-Werror,-Wrange-loop-analysis]
    for(const auto succ : succs)`


CBMC version: 5.14.1 (cbmc-5.14.1-dirty)
Operating system: MacOS Catalina 10.15.6
Exact command line resulting in the issue: cmake --build build
What behaviour did you expect: building without errors
What happened instead: build fails

--- a/src/goto-programs/goto_program.cpp
+++ b/src/goto-programs/goto_program.cpp
@@ -1093,7 +1093,7 @@ bool goto_programt::equals(const goto_programt &other) const
 
     // the number of targets is the same as instructiont::equals returned "true"
     auto other_target_it = other_it->targets.begin();
-    for(const auto t : ins.targets)
+    for(const auto &t : ins.targets)
     {
       if(
         t->location_number - ins.location_number !=

--- a/src/goto-symex/build_goto_trace.cpp
+++ b/src/goto-symex/build_goto_trace.cpp
@@ -329,7 +329,7 @@ void build_goto_trace(
 
   for(const auto &time_and_ssa_steps : time_map)
   {
-    for(const auto ssa_step_it : time_and_ssa_steps.second)
+    for(const auto &ssa_step_it : time_and_ssa_steps.second)
     {
       const auto &SSA_step = *ssa_step_it;
       goto_trace.steps.push_back(goto_trace_stept());

--- a/src/solvers/sat/cnf.cpp
+++ b/src/solvers/sat/cnf.cpp
@@ -166,7 +166,7 @@ literalt cnft::land(const bvt &bv)
   if(bv.size()==2)
     return land(bv[0], bv[1]);
 
-  for(const auto l : bv)
+  for(const auto &l : bv)
     if(l.is_false())
       return l;
 
@@ -179,7 +179,7 @@ literalt cnft::land(const bvt &bv)
   literalt literal=new_variable();
   lits[1]=neg(literal);
 
-  for(const auto l : new_bv)
+  for(const auto &l : new_bv)
   {
     lits[0]=pos(l);
     lcnf(lits);
@@ -188,7 +188,7 @@ literalt cnft::land(const bvt &bv)
   lits.clear();
   lits.reserve(new_bv.size()+1);
 
-  for(const auto l : new_bv)
+  for(const auto &l : new_bv)
     lits.push_back(neg(l));
 
   lits.push_back(pos(literal));
@@ -209,7 +209,7 @@ literalt cnft::lor(const bvt &bv)
   if(bv.size()==2)
     return lor(bv[0], bv[1]);
 
-  for(const auto l : bv)
+  for(const auto &l : bv)
     if(l.is_true())
       return l;
 
@@ -222,7 +222,7 @@ literalt cnft::lor(const bvt &bv)
   literalt literal=new_variable();
   lits[1]=pos(literal);
 
-  for(const auto l : new_bv)
+  for(const auto &l : new_bv)
   {
     lits[0]=neg(l);
     lcnf(lits);
@@ -231,7 +231,7 @@ literalt cnft::lor(const bvt &bv)
   lits.clear();
   lits.reserve(new_bv.size()+1);
 
-  for(const auto l : new_bv)
+  for(const auto &l : new_bv)
     lits.push_back(pos(l));
 
   lits.push_back(neg(literal));
@@ -254,7 +254,7 @@ literalt cnft::lxor(const bvt &bv)
 
   literalt literal=const_literal(false);
 
-  for(const auto l : bv)
+  for(const auto &l : bv)
     literal=lxor(l, literal);
 
   return literal;
@@ -404,7 +404,7 @@ bvt cnft::eliminate_duplicates(const bvt &bv)
   bvt dest;
   dest.reserve(bv.size());
 
-  for(const auto l : bv)
+  for(const auto &l : bv)
     if(s.insert(l).second)
       dest.push_back(l);
 
@@ -423,7 +423,7 @@ bool cnft::process_clause(const bvt &bv, bvt &dest)
 
   // first check simple things
 
-  for(const auto l : bv)
+  for(const auto &l : bv)
   {
     // we never use index 0
     INVARIANT(l.var_no() != 0, "variable 0 must not be used");
@@ -449,7 +449,7 @@ bool cnft::process_clause(const bvt &bv, bvt &dest)
   dest.clear();
   dest.reserve(bv.size());
 
-  for(const auto l : bv)
+  for(const auto &l : bv)
   {
     if(l.is_false())
       continue; // remove

--- a/src/solvers/solver_hardness.cpp
+++ b/src/solvers/solver_hardness.cpp
@@ -165,7 +165,7 @@ void solver_hardnesst::produce_report()
       json_stringt{assertion_stats.ssa_expression};
 
     auto assertion_stats_pcs_json = json_arrayt{};
-    for(const auto pc : assertion_stats.pcs)
+    for(const auto &pc : assertion_stats.pcs)
     {
       auto assertion_stats_pc_json = json_objectt{};
       assertion_stats_pc_json["GOTO"] =

--- a/src/util/graph.h
+++ b/src/util/graph.h
@@ -632,7 +632,7 @@ std::vector<typename N::node_indext> grapht<N>::get_reachable(
 
     const auto &node = nodes[n];
     const auto &succs = forwards ? node.out : node.in;
-    for(const auto succ : succs)
+    for(const auto &succ : succs)
       if(!visited[succ.first])
         s.push(succ.first);
   }

--- a/src/util/sharing_map.h
+++ b/src/util/sharing_map.h
@@ -619,7 +619,7 @@ SHARING_MAPT(void)
       const to_mapt &m = ip->get_to_map();
       SM_ASSERT(!m.empty());
 
-      for(const auto &item : m)
+      for(const auto item : m)
       {
         stack.push(&item.second);
       }
@@ -694,7 +694,7 @@ SHARING_MAPT(std::size_t)
       const to_mapt &m = ip->get_to_map();
       SM_ASSERT(!m.empty());
 
-      for(const auto &item : m)
+      for(const auto item : m)
       {
         stack.push(&item.second);
       }
@@ -939,7 +939,7 @@ SHARING_MAPT(void)
 
       if(ip2->is_internal())
       {
-        for(const auto &item : ip1->get_to_map())
+        for(const auto item : ip1->get_to_map())
         {
           const nodet &child = item.second;
 
@@ -964,7 +964,7 @@ SHARING_MAPT(void)
       {
         SM_ASSERT(ip2->is_leaf());
 
-        for(const auto &item : ip1->get_to_map())
+        for(const auto item : ip1->get_to_map())
         {
           const nodet &child = item.second;
 

